In [178]:
# Import Libraries

import pandas as pd
import numpy as np
import datetime
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [179]:
#read CSV File

Year_2019 = pd.read_csv("D:\\GA Work_Essia\\Data\\metro-bike-share-trips-2019-q4.csv")
Year_2019.head()

trip_id  duration           start_time             end_time  \
0  130053086        25  2019-10-01 00:03:10  2019-10-01 00:28:40   
1  130053085         2  2019-10-01 00:07:43  2019-10-01 00:09:44   
2  130053084        15  2019-10-01 00:17:30  2019-10-01 00:32:01   
3  130053083         4  2019-10-01 00:18:20  2019-10-01 00:22:34   
4  130053082        20  2019-10-01 00:22:09  2019-10-01 00:42:53   

   start_station  start_lat   start_lon  end_station    end_lat     end_lon  \
0           3049  34.056969 -118.253593         3038  34.046822 -118.248352   
1           3067  34.045181 -118.250237         3029  34.048851 -118.246422   
2           4420  34.168629 -118.377068         4416  34.157581 -118.388260   
3           3031  34.044701 -118.252441         3006  34.045540 -118.256668   
4           3042  34.049301 -118.238808         4314  34.057709 -118.279762   

   bike_id  plan_duration trip_route_category passholder_type bike_type  
0    16580            365             One Way     Annual Pass  electric  
1    19587             30             One Way    Monthly Pass  electric  
2    15625             30             One Way    Monthly Pass     smart  
3    12450            365             One Way     Annual Pass  standard  
4    17519            365             One Way     Annual Pass  electric

In [180]:
#Change datatype
    
Year_2019["start_time"]=pd.to_datetime(Year_2019['start_time'])
Year_2019["end_time"]=pd.to_datetime(Year_2019['end_time'])

In [181]:
#extract date from time

Year_2019["start_day"]=Year_2019["start_time"].dt.date
Year_2019["end_day"]=Year_2019["end_time"].dt.date

#extract hour from time

Year_2019["start_hour"]=Year_2019["start_time"].dt.hour
Year_2019["end_hour"]=Year_2019["end_time"].dt.hour
# Find day name

Year_2019["weekday"]=Year_2019["start_time"].dt.weekday_name



In [182]:
Year_2019_Q4=Year_2019[['start_station','start_day','start_hour','weekday','end_station','end_day','end_hour','duration','bike_id','passholder_type','plan_duration','trip_route_category','bike_type']]
Year_2019_Q4.head()

start_station   start_day  start_hour  weekday  end_station     end_day  \
0           3049  2019-10-01           0  Tuesday         3038  2019-10-01   
1           3067  2019-10-01           0  Tuesday         3029  2019-10-01   
2           4420  2019-10-01           0  Tuesday         4416  2019-10-01   
3           3031  2019-10-01           0  Tuesday         3006  2019-10-01   
4           3042  2019-10-01           0  Tuesday         4314  2019-10-01   

   end_hour  duration  bike_id passholder_type  plan_duration  \
0         0        25    16580     Annual Pass            365   
1         0         2    19587    Monthly Pass             30   
2         0        15    15625    Monthly Pass             30   
3         0         4    12450     Annual Pass            365   
4         0        20    17519     Annual Pass            365   

  trip_route_category bike_type  
0             One Way  electric  
1             One Way  electric  
2             One Way     smart  
3             One Way  standard  
4             One Way  electric

In [183]:
Year_2019_Q4['start_day'].value_counts().head(1)

2019-10-06    2323
Name: start_day, dtype: int64

In [184]:
Year_2019_Q4["start_day"]=pd.to_datetime(Year_2019_Q4['start_day'])

C:\Users\Swap\Anaconda\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [185]:
data_1day=Year_2019_Q4[(Year_2019_Q4['start_day'] > '2019-10-05') & (Year_2019_Q4['start_day'] < '2019-10-07')]
data_1day=data_1day[(data_1day['start_hour'] > 7) & (data_1day['start_hour'] < 21)]

In [186]:
data_1day['start_station'].value_counts().head()

3005    217
3030    109
3069     90
3042     89
4314     88
Name: start_station, dtype: int64

3005 is busy station on 2019-10-06 Station name-7th & Flower
whereas station near to Union staion is 3014 which as 44 entries.

### Smart Bike available only in few stations and no sttaion has all three bikes together

In [187]:
data_3005_start=data_1day[(data_1day['start_station'] > 3004) & (data_1day['start_station'] < 3006)]
data_3005_end=data_1day[(data_1day['end_station'] > 3004) & (data_1day['end_station'] < 3006)]

In [188]:
data_3005_start["bike_Left_count"] = data_3005_start["bike_id"].count()
data_3005_end["bike_Arrive_count"] = data_3005_end["bike_id"].count()

C:\Users\Swap\Anaconda\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\Swap\Anaconda\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [200]:
Left=pd.pivot_table(data_3005_start,index=["start_station","start_hour"],columns=["bike_type"],values=["bike_Left_count"],aggfunc="count",margins=True)
Left

bike_Left_count              
bike_type                       electric standard  All
start_station start_hour                              
3005          8                      2.0      NaN    2
              9                      3.0      8.0   11
              10                     3.0     25.0   28
              11                     8.0     35.0   43
              12                     6.0     29.0   35
              13                     4.0     29.0   33
              14                     1.0     23.0   24
              15                     8.0      8.0   16
              16                     3.0      7.0   10
              17                     3.0      NaN    3
              18                     3.0      3.0    6
              19                     2.0      1.0    3
              20                     3.0      NaN    3
All                                 49.0    168.0  217

In [190]:
Left.columns = ["_".join((j,k)) for j,k in Left.columns]
Left.reset_index()

start_station start_hour  bike_Left_count_electric  \
0           3005          8                       2.0   
1           3005          9                       3.0   
2           3005         10                       3.0   
3           3005         11                       8.0   
4           3005         12                       6.0   
5           3005         13                       4.0   
6           3005         14                       1.0   
7           3005         15                       8.0   
8           3005         16                       3.0   
9           3005         17                       3.0   
10          3005         18                       3.0   
11          3005         19                       2.0   
12          3005         20                       3.0   
13           All                                 49.0   

    bike_Left_count_standard  bike_Left_count_All  
0                        NaN                    2  
1                        8.0                   11  
2                       25.0                   28  
3                       35.0                   43  
4                       29.0                   35  
5                       29.0                   33  
6                       23.0                   24  
7                        8.0                   16  
8                        7.0                   10  
9                        NaN                    3  
10                       3.0                    6  
11                       1.0                    3  
12                       NaN                    3  
13                     168.0                  217

In [201]:
Arrive=pd.pivot_table(data_3005_end,index=["end_station","end_hour"],columns=["bike_type"],values=["bike_Arrive_count"],aggfunc="count",margins=True)
print(Arrive)

                     bike_Arrive_count              
bike_type                     electric standard  All
end_station end_hour                                
3005        8                      2.0      1.0    3
            9                      2.0      NaN    2
            10                     4.0      4.0    8
            11                     7.0     18.0   25
            12                     5.0     24.0   29
            13                     4.0     15.0   19
            14                     6.0     20.0   26
            15                     5.0     27.0   32
            16                     4.0      7.0   11
            17                     1.0      1.0    2
            18                     4.0      NaN    4
            19                     2.0      1.0    3
            20                     4.0      NaN    4
All                               50.0    118.0  168


In [192]:
Arrive.columns = ["_".join((j,k)) for j,k in Arrive.columns]
Arrive.reset_index()

end_station end_hour  bike_Arrive_count_electric  \
0         3005        8                         2.0   
1         3005        9                         2.0   
2         3005       10                         4.0   
3         3005       11                         7.0   
4         3005       12                         5.0   
5         3005       13                         4.0   
6         3005       14                         6.0   
7         3005       15                         5.0   
8         3005       16                         4.0   
9         3005       17                         1.0   
10        3005       18                         4.0   
11        3005       19                         2.0   
12        3005       20                         4.0   
13         All                                 50.0   

    bike_Arrive_count_standard  bike_Arrive_count_All  
0                          1.0                      3  
1                          NaN                      2  
2                          4.0                      8  
3                         18.0                     25  
4                         24.0                     29  
5                         15.0                     19  
6                         20.0                     26  
7                         27.0                     32  
8                          7.0                     11  
9                          1.0                      2  
10                         NaN                      4  
11                         1.0                      3  
12                         NaN                      4  
13                       118.0                    168

In [202]:
result=pd.concat([Left,Arrive],axis=1)
result
result.reset_index()

start_station start_hour bike_Left_count                \
bike_type                                 electric standard  All   
0                  3005          8             2.0      NaN    2   
1                  3005          9             3.0      8.0   11   
2                  3005         10             3.0     25.0   28   
3                  3005         11             8.0     35.0   43   
4                  3005         12             6.0     29.0   35   
5                  3005         13             4.0     29.0   33   
6                  3005         14             1.0     23.0   24   
7                  3005         15             8.0      8.0   16   
8                  3005         16             3.0      7.0   10   
9                  3005         17             3.0      NaN    3   
10                 3005         18             3.0      3.0    6   
11                 3005         19             2.0      1.0    3   
12                 3005         20             3.0      NaN    3   
13                  All                       49.0    168.0  217   

          bike_Arrive_count                
bike_type          electric standard  All  
0                       2.0      1.0    3  
1                       2.0      NaN    2  
2                       4.0      4.0    8  
3                       7.0     18.0   25  
4                       5.0     24.0   29  
5                       4.0     15.0   19  
6                       6.0     20.0   26  
7                       5.0     27.0   32  
8                       4.0      7.0   11  
9                       1.0      1.0    2  
10                      4.0      NaN    4  
11                      2.0      1.0    3  
12                      4.0      NaN    4  
13                     50.0    118.0  168

In [203]:
result.index.names

FrozenList(['start_station', 'start_hour'])

In [204]:
result.index.names = ['Station_ID', 'Hour']

In [205]:
result

bike_Left_count               bike_Arrive_count              
bike_type              electric standard  All          electric standard  All
Station_ID Hour                                                              
3005       8                2.0      NaN    2               2.0      1.0    3
           9                3.0      8.0   11               2.0      NaN    2
           10               3.0     25.0   28               4.0      4.0    8
           11               8.0     35.0   43               7.0     18.0   25
           12               6.0     29.0   35               5.0     24.0   29
           13               4.0     29.0   33               4.0     15.0   19
           14               1.0     23.0   24               6.0     20.0   26
           15               8.0      8.0   16               5.0     27.0   32
           16               3.0      7.0   10               4.0      7.0   11
           17               3.0      NaN    3               1.0      1.0    2
           18               3.0      3.0    6               4.0      NaN    4
           19               2.0      1.0    3               2.0      1.0    3
           20               3.0      NaN    3               4.0      NaN    4
All                        49.0    168.0  217              50.0    118.0  168